core/fault/stacktrace/pack.go

```diff
@@ -0,0 +1,374 @@
+		v := val ^ prev
+		out[i], prev = v, v
+	}
+	return out
+}
+
+type backRefCompressor struct {
+	pack    func(ref, val uint64) uint64
+	unpack  func(ref, packed uint64) uint64
+	history int
+}
+
+func packXor(ref, val uint64) uint64      { return ref ^ val }
+func unpackXor(ref, packed uint64) uint64 { return ref ^ packed }
+
+func packDiff(ref, val uint64) uint64 {
+	if ref > val {
+		return (ref - val) << 1
+	}
+	return (val-ref)<<1 | 1
+}
+func unpackDiff(ref, val uint64) uint64 {
+	sign, val := val&1, val>>1
+	if sign == 0 {
+		return ref + val
+	}
+	return ref - val
+}
+
+func (p backRefCompressor) compress(c []uint64) []uint64 {
+	out := make([]uint64, 0, len(c))
+
+	size := func(val uint64) uint32 {
+		bs := binary.BitStream{}
+		writeVLE(val, &bs)
+		return bs.WritePos
+	}
+
+	for i, v := range c {
+		bestValue := v
+		bestIndex := 0
+		bestScore := size(uint64(bestIndex)) + size(bestValue)
+		for idx := 1; idx < p.history; idx++ {
+			absIdx := i - idx
+			if absIdx < 0 {
+				continue
+			}
+			val := p.pack(v, c[absIdx])
+			if score := size(val) + size(uint64(idx)); score < bestScore {
+				bestValue, bestIndex, bestScore = val, idx, score
+			}
+		}
+
+		out = append(out, uint64(bestIndex), bestValue)
+
+		if debug {
+			printVLE(v)
+			fmt.Printf("(%d) -- <%v> ", size(v), bestIndex)
+			printVLE(uint64(bestIndex))
+			fmt.Printf(", ")
+			printVLE(bestValue)
+			fmt.Printf("(%d) ", bestScore)
+			fmt.Printf("\n")
+		}
+	}
+	return out
+}
+
+func (p backRefCompressor) decompress(c []uint64) []uint64 {
+	out := make([]uint64, 0, len(c)/2)
+	for len(c) >= 2 {
+		idx, val := int(c[0]), c[1]
+		if idx == 0 {
+			out = append(out, val)
+		} else {
+			ref := out[len(out)-idx]
+			out = append(out, p.unpack(ref, val))
+		}
+		c = c[2:]
+	}
+	return out
+}
+
+type dictionaryCompressor struct{}
+
+func (p dictionaryCompressor) compress(c []uint64) []uint64 {
+	unique := map[uint64]struct{}{}
+	for _, v := range c {
+		unique[v] = struct{}{}
+	}
+
+	sorted := make([]uint64, 0, len(unique))
+	for v := range unique {
+		sorted = append(sorted, v)
+	}
+	sort.Slice(sorted, func(i, j int) bool { return sorted[i] < sorted[j] })
+
+	indices := map[uint64]uint64{}
+	for i, v := range sorted {
+		indices[v] = uint64(i)
+	}
+
+	out := make([]uint64, 0, len(c)*2)
+
+	// Delta
+	prev := uint64(0)
+	for _, v := range sorted {
+		delta := v - prev
+		out, prev = append(out, delta), v
+	}
+
+	// Delimiter
+	out = append(out, 0)
+
+	// Values
+	for _, v := range c {
+		idx := indices[v]
+		out = append(out, uint64(idx))
+	}
+
+	return out
+}
+
+func (p dictionaryCompressor) decompress(c []uint64) []uint64 {
+	table, prev := make([]uint64, 0, len(c)), uint64(0)
+	for i, delta := range c {
+		if delta == 0 {
+			c = c[i+1:]
+			break
+		}
+		val := prev + delta
+		table, prev = append(table, val), val
+	}
+
+	out := make([]uint64, len(c))
+	for i, v := range c {
+		out[i] = table[v]
+	}
+	return out
+}
+
+type compressorList []compressor
```

Comment:
That seems like unnecessary complexity.  Does that gain anything at all?

Reply:
Yeah - it lets you combine compressors together. Experimenting with different compressor combinations has been really useful.